acceptancetests/assess_network_spaces.py

```diff
@@ -0,0 +1,600 @@
+            "172.31.0.0/20": {
+                "provider-id": "subnet-38f9d07e",
+                "provider-network-id": "vpc-1f40b47a",
+                "space": "",
+                "status": "in-use",
+                "type": "ipv4",
+                "zones": [
+                    "us-east-1a"
+                ]
+             }
+         }
+    """
+    newsubnets = {}
+    if 'subnets' in subnets:
+        newsubnets['subnets'] = {}
+        for subnet in subnets['subnets'].keys():
+            if 'INFAN' not in subnets['subnets'][subnet]['provider-id']:
+                newsubnets['subnets'][subnet] = subnets['subnets'][subnet]
+
+    """ Example dict output from juju list-spaces:
+        "spaces": {
+            "space1": {
+                "172.31.16.0/20": {
+                    "provider-id": "subnet-13a6aa67",
+                    "status": "in-use",
+                    "type": "ipv4",
+                    "zones": [
+                        "us-east=1d"
+                    ]
+                }
+            }
+        }
+    """
+    if 'spaces' in subnets:
+        newsubnets['spaces'] = {}
+        for space in subnets['spaces'].keys():
+            for subnet in subnets['spaces'][space].keys():
+                if ('INFAN' not in
+                        subnets['spaces'][space][subnet]['provider-id']):
+                    newsubnets['spaces'].setdefault(space, {})
+                    newsubnets['spaces'][space][subnet] = \
+                        subnets['spaces'][space][subnet]
+
+    return newsubnets
+
+
+def machine_can_ping_ip(client, machine, ip):
+    """SSH to the machine and attempt to ping the given IP.
+
+    :param client: juju client object
+    :param machine: machine to connect to
+    :param ip: IP address to ping
+    :returns: success of ping
+    """
+    rc, _ = client.juju('ssh',
+                ('--proxy', machine, 'ping -c1 -q ' + ip), check=False)
+    return rc == 0
+
+
+def ip_in_cidr(address, cidr):
+    """Returns true if the ip address given is within the range defined
+    by the cidr subnet.
+
+    :param address: A valid IPv4 address
+    :param cidr: A valid subnet in CIDR notation
+    """
+    return (ipaddress.ip_address(address.decode('utf-8'))
+            in ipaddress.ip_network(cidr.decode('utf-8')))
+
+
+def parse_args(argv):
+    """Parse all arguments."""
+    parser = argparse.ArgumentParser(description="Test Network Spaces")
+    add_basic_testing_arguments(parser)
+    parser.add_argument('--model', help='Existing Juju model to test against')
+    parser.set_defaults(series='xenial')
+    return parser.parse_args(argv)
+
+
+def start_test(client, args):
+    """Launch the test and perform some cleanup."""
+    test = AssessNetworkSpaces(args)
+    try:
+        test.assess_network_spaces(client, args.model, args.series)
+    finally:
+        if args.model:
+            test.cleanup(client)
+            log.info('Cleanup complete.')
+
+def vpc_bootstrap(args):
```

Comment:
I think I see where you're going here. You're saying we create a class, VPCAws that is a subclass of VPC so that we can add other subclasses later as spaces is implemented in other providers? I think the nomenclature is not quite right since VPC is an amazon thing. If we are to abstract this to other providers, the thing that is different is how spaces are set up.

Currently, two providers are supported, AWS (ec2) and MaaS. I don't know enough about how spaces under MaaS is different to know how much needs to be implemented differently.

Maybe instead of calling it VPC, we call it Spaces and we create subclasses SpacesAWS and SpacesMaaS. We'd have methods for pre_bootstrap and cleanup and maybe setup_spaces which would get called after we've bootstrapped but before running tests. Instead of exposing vpcid (which would not exist for MaaS), the pre_bootstrap method should take a bs_manager object as a parameter. The pre_bootstrap method would take care of feeding the vpcid to the bs_manager's client environment or setting up the MaaS excluded and reserved spaces.